internal/engine/wazevo/backend/isa/amd64/lower_mem.go

```diff
@@ -0,0 +1,164 @@
+			// Absorb the offset into the amode with no index.
+			am = newAmodeImmReg(uint32(offset), base)
+			offset = 0
+		} else {
+			// Offset is too large to be absorbed into the amode, will be added later.
+			am = newAmodeImmReg(0, base)
+		}
+	} else if index := a64s.dequeue(); lower32willSignExtendTo64(uint64(offset)) {
+		// Absorb the offset into the amode with an index.
+		am = newAmodeRegRegShift(uint32(offset), base, index, 0)
+		offset = 0
+	} else {
+		// Offset is too large to be absorbed into the amode, will be added later.
+		am = newAmodeRegRegShift(0, base, index, 0)
+	}
+
+	baseReg := am.base
+	if offset > 0 {
+		baseReg = m.addConstToReg64(baseReg, offset) // baseReg += offset
+	}
+
+	for !a64s.empty() {
+		a64 := a64s.dequeue()
+		baseReg = m.addReg64ToReg64(baseReg, a64) // baseReg += a64
+	}
+
+	am.base = baseReg
+	return
+}
+
+var addendsMatchOpcodes = [4]ssa.Opcode{ssa.OpcodeUExtend, ssa.OpcodeSExtend, ssa.OpcodeIadd, ssa.OpcodeIconst}
+
+func (m *machine) collectAddends(ptr ssa.Value) (addends64 *queue[regalloc.VReg], offset int64) {
+	m.addendsWorkQueue.reset()
+	m.addends64.reset()
+	m.addendsWorkQueue.enqueue(ptr)
+
+	for !m.addendsWorkQueue.empty() {
+		v := m.addendsWorkQueue.dequeue()
+
+		def := m.c.ValueDefinition(v)
+		switch op := m.c.MatchInstrOneOf(def, addendsMatchOpcodes[:]); op {
+		case ssa.OpcodeIadd:
+			// If the addend is an add, we recursively collect its operands.
+			x, y := def.Instr.Arg2()
+			m.addendsWorkQueue.enqueue(x)
+			m.addendsWorkQueue.enqueue(y)
+			def.Instr.MarkLowered()
+		case ssa.OpcodeIconst:
+			// If the addend is constant, we just statically merge it into the offset.
+			ic := def.Instr
+			u64 := ic.ConstantVal()
+			if ic.Return().Type().Bits() == 32 {
+				offset += int64(int32(u64)) // sign-extend.
+			} else {
+				offset += int64(u64)
+			}
+			def.Instr.MarkLowered()
+		case ssa.OpcodeUExtend, ssa.OpcodeSExtend:
+			switch input := def.Instr.Arg(); input.Type().Bits() {
+			case 64:
+				// If the input is already 64-bit, this extend is a no-op. TODO: shouldn't this be optimized out at much earlier stage? no?
+				m.addends64.enqueue(m.getOperand_Reg(m.c.ValueDefinition(input)).r)
+				def.Instr.MarkLowered()
+				continue
+			case 32:
+				inputDef := m.c.ValueDefinition(input)
+				constInst := inputDef.IsFromInstr() && inputDef.Instr.Constant()
+				switch {
+				case constInst && op == ssa.OpcodeUExtend:
+					// Zero-extension of a 32-bit constant can be merged into the offset.
+					offset += int64(uint32(inputDef.Instr.ConstantVal()))
+				case constInst && op == ssa.OpcodeSExtend:
+					// Sign-extension of a 32-bit constant can be merged into the offset.
+					offset += int64(int32(inputDef.Instr.ConstantVal())) // sign-extend!
+				default:
+					// Cannot fold into a constant, ignore.
+					continue
+				}
+				def.Instr.MarkLowered()
+				continue
+			}
+			// Note: case Ishl x, y could be handled too when the offset amount is <= 3.
```

Comment:
I am trying to do this; basically, instead of returning VReg we can return a pair `(VReg, byte)`, but then folding complicates a bit; e.g. when a single `(VReg, byte)` is present, we should either use the missing `amodeImmRegShift` or zero-out some register so that we can use `amodeRegRegImmShift`.

Even when we `add` together all the remaining addends extra care becomes necessary because in some cases we might want to substitute `(add x (shl y amt))` with `(add x [y+amt])`; but even in this case `[y+amt]` is either  `amodeRegRegImmShift` or needs a zero reg.

Reply:
I have added this to commit 6c687e40 but we can revert it if that adds too much noise